### PR TITLE
[CUDA graphs][BC-breaking] Removes post-backward syncs on default stream

### PR DESCRIPTION
**Pull request by mcarilli**
Before https://github.com/pytorch/pytorch/pull/57833, calls to backward() or grad() synced only the calling thread's default stream with autograd leaf streams at the end of backward. This made the following weird pattern safe:
```python
with torch.cuda.stream(s):
    # imagine forward used many streams, so backward leaf nodes may run on many streams
    loss.backward()
# no sync
use grads
```

but a more benign-looking pattern was unsafe:
```python
with torch.cuda.stream(s):
    # imagine forward used a lot of streams, so backward leaf nodes may run on many streams
    loss.backward()
    # backward() syncs the default stream with all the leaf streams, but does not sync s with anything,
    # so counterintuitively (even though we're in the same stream context as backward()!)
    # it is NOT SAFE to use grads here, and there's no easy way to make it safe,
    # unless you manually sync on all the streams you used in forward,
    # or move "use grads" back to default stream outside the context.
    use grads
```
@mruberry @ngimel and I decided backward() should have the [same user-facing stream semantics as any cuda op](https://pytorch.org/docs/master/notes/cuda.html#stream-semantics-of-backward-passes).** In other words, the weird pattern should be unsafe, and the benign-looking pattern should be safe. Implementationwise, this meant backward() should sync its calling thread's current stream, not default stream, with the leaf streams.

After https://github.com/pytorch/pytorch/pull/57833, backward syncs the calling thread's current stream AND default stream with all leaf streams at the end of backward. The default stream syncs were retained for temporary backward compatibility.

This PR finishes https://github.com/pytorch/pytorch/pull/57833's work by deleting syncs on the default stream.

With this PR, graph-capturing an entire backward() call should be possible (see the [test_graph_grad_scaling diffs](https://github.com/pytorch/pytorch/compare/master...mcarilli:streaming_backwards_remove_default_syncs?expand=1#diff-893b1eea27352f336f4cd832919e48d721e4e90186e63400b8596db6b82e7450R3641-R3642)).

** first paragraph has a formatting error which this PR should also fix.

### CODE REVIEW

File: docs/source/notes/cuda.rst [hunk 1]
```diff
@@ -255,12 +259,6 @@ have the same stream-semantics relationship as any group of ops::
         initial_grad.record_stream(s)
         loss.backward(gradient=initial_grad)
 
-If your forward pass runs some independent ops in parallel on different streams,
```
Comment (mruberry) [inline docs/source/notes/cuda.rst:258]: What about adding a sentence or two here that in prior versions of PyTorch (1.9 and below) these operations synced with the default CUDA stream, but this is no longer the case. Then you could suggest how people could restore the old behavior by calling `torch.cuda.synchronize()` if they were concerned?

Comment (mcarilli) [reply docs/source/notes/cuda.rst:258]: Added https://github.com/pytorch/pytorch/pull/60421/files#diff-4840823ff4b0cd7fe61b9ada2a1746f6936c944bfbcc4d1bdb8b3dc1561d4401R262-R281

**Review by ngimel — APPROVED** [review]

File: docs/source/notes/cuda.rst [hunk 2]
```diff
@@ -255,11 +259,26 @@ have the same stream-semantics relationship as any group of ops::
...
-A. helps the backward pass exploit that same parallelism.
+BC note: Using grads on the default stream
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In prior versions of Pytorch (1.9 and earlier), the autograd engine always synced
+the default stream with all backward ops, so the following pattern:
```
Comment (ngimel) [inline docs/source/notes/cuda.rst:266]:
```suggestion
the default stream with all backward ops, so the following pattern::
```

File: docs/source/notes/cuda.rst [hunk 3]
```diff
@@ -255,11 +259,26 @@ have the same stream-semantics relationship as any group of ops::
...
+
+    with torch.cuda.stream(s):
+        loss.backward()
+    use grads
+
+was safe as long as ``use grads`` happened on the default stream.
+In present Pytorch, that pattern is no longer safe. If ``backward()``
+and ``use grads`` are in different stream contexts, you must sync the streams:
```
Comment (ngimel) [inline docs/source/notes/cuda.rst:274]:
```suggestion
and ``use grads`` are in different stream contexts, you must sync the streams::
```

**Pull request closed without merging**